velox/common/base/Status.h

```diff
@@ -0,0 +1,459 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Adapted from Apache Arrow.
+
+#pragma once
+
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <folly/Likely.h>
+#include <string>
+#include <utility>
+
+namespace facebook::velox {
+
+/// The Status object is an object holding the outcome of an operation (success
+/// or error).
+///
+/// The outcome is represented as a StatusCode, holding either a success
+/// (StatusCode::kOK) or an error (any other of the StatusCode enumeration
+/// values). If an error occurred, a specific error message is generally
+/// attached.
+///
+/// The status object is commonly allocated in the stack, so it needs be compact
+/// to be efficient. For the common success case, its size is a single (nullptr)
+/// pointer. For failure cases, it allocates an external object containing the
+/// StatusCode and error message. This keeps the object compact and prevents
+/// allocations in the common success case. The same strategy is used in Status
+/// object from other well-known libraries like Apache Arrow, RocksDB, Kudu, and
+/// others.
+///
+/// Simple usage:
+///
+///  Status operation() {
+///    if (noMoreMemory) {
+///      return Status::OutOfMemory("Not enough memory to run 'operation'!");
+///    }
+///    return Status::OK();
+///  }
+///
+/// Call site:
+///
+///  auto status = operation();
+///  if (status.ok()) {
+///    ...
+///  } else if (status.isOutOfMemory()) {
+///    (gracefully handle out of memory)
+///  } else if (status.isNotImplemented()) {
+///    (and so on)
+///  }
+///
+/// Other common usage patterns:
+///
+/// The same logic above can be implemented using helper macros:
+///
+///  Status operation() {
+///    VELOX_RETURN_IF(noMoreMemory, Status::OutOfMemory(
+///       "Not enough memory to run 'operation'!"));
+///    ...
+///    return Status::OK();
+///  }
+///
+/// To ensure operations succeed (or if not, return the same status from the
+/// current function):
+///
+///  ...
+///  VELOX_RETURN_NOT_OK(operation1());
+///  VELOX_RETURN_NOT_OK(operation2());
+///  VELOX_RETURN_NOT_OK(operation3());
+///  ...
+
+enum class StatusCode : int8_t {
+  kOK = 0,
+  kUserError = 1,
```

Comment:
It would be nice to document these errors codes and provide some guidance about adding more.